index.js

```diff
@@ -10,146 +10,137 @@ var curve = ecurve.getCurveByName('secp256k1')
-  var d = cs.decode(wif).slice(1)
-  var compressed = (d.length === 33) && (d[32] === 0x01)
-
-  // truncate the compression flag
-  if (compressed) {
-    d = d.slice(0, -1)
-  }
-
-  return cs.encode(this.encryptRaw(d, compressed, passphrase, saltAddress, progressCallback))
+function encrypt (buffer, compressed, passphrase, address, progressCallback) {
+  return bs58check.encode(encryptRaw(buffer, compressed, passphrase, address, progressCallback))
 }
 
 // some of the techniques borrowed from: https://github.com/pointbiz/bitaddress.org
 // todo: (optimization) init buffer in advance, and use copy instead of concat
-Bip38.prototype.decryptRaw = function (encData, passphrase, progressCallback) {
+function decryptRaw (buffer, passphrase, progressCallback) {
   // 39 bytes: 2 bytes prefix, 37 bytes payload
-  assert.equal(encData.length, 39, 'Invalid BIP38 data length')
-
-  // first byte is always 0x01
-  assert.equal(encData.readUInt8(0), 0x01, 'Invalid BIP38 prefix')
+  if (buffer.length !== 39) throw new Error('Invalid BIP38 data length')
+  if (buffer.readUInt8(0) !== 0x01) throw new Error('Invalid BIP38 prefix')
 
   // check if BIP38 EC multiply
-  var type = encData.readUInt8(1)
-  if (type === 0x43) {
-    return this.decryptECMult(encData, passphrase, progressCallback)
-  }
+  var type = buffer.readUInt8(1)
+  if (type === 0x43) return decryptECMult(buffer, passphrase, progressCallback)
+  if (type !== 0x42) throw new Error('Invalid BIP38 type')
 
   passphrase = new Buffer(passphrase, 'utf8')
 
-  assert.equal(type, 0x42, 'Invalid BIP38 type')
-  var flagByte = encData.readUInt8(2)
+  var flagByte = buffer.readUInt8(2)
   var compressed = flagByte === 0xe0
+  if (!compressed && flagByte !== 0xc0) throw new Error('Invalid BIP38 compression flag')
 
-  if (!compressed) {
-    assert.equal(flagByte, 0xc0, 'Invalid BIP38 compression flag')
-  }
-
-  var N = this.scryptParams.N
-  var r = this.scryptParams.r
-  var p = this.scryptParams.p
+  var N = SCRYPT_PARAMS.N
+  var r = SCRYPT_PARAMS.r
+  var p = SCRYPT_PARAMS.p
 
-  var addresshash = encData.slice(3, 7)
-  var scryptBuf = scrypt(passphrase, addresshash, N, r, p, 64, progressCallback)
+  var salt = buffer.slice(3, 7)
+  var scryptBuf = scrypt(passphrase, salt, N, r, p, 64, progressCallback)
   var derivedHalf1 = scryptBuf.slice(0, 32)
   var derivedHalf2 = scryptBuf.slice(32, 64)
 
-  var privKeyBuf = encData.slice(7, 7 + 32)
-  var decipher = aes.createDecipheriv('aes-256-ecb', derivedHalf2, new Buffer(0))
+  var privKeyBuf = buffer.slice(7, 7 + 32)
+  var decipher = aes.createDecipheriv('aes-256-ecb', derivedHalf2, NULL)
   decipher.setAutoPadding(false)
   decipher.end(privKeyBuf)
 
   var plainText = decipher.read()
   var privateKey = xor(plainText, derivedHalf1)
 
+  // verify salt matches address
+//   var d = BigInteger.fromBuffer(privateKey)
+//   var address = getAddress(d)
+//   var checksum = hash256(address).slice(0, 4)
+//   assert.deepEqual(salt, checksum)
```

Comment:
Left out for now, as it instantly fails the tests... I'm not sure why.  But it is **unrelated** to this PR,  as the tests failed previously if this was added.

Lets get to that immediately after this :+1: